Putting in the design for mapping remote work items

sbose78 opened the pull request:
As per discussion and design suggestion by @aslakknutsen on https://github.com/almighty/almighty-core/issues/131  

Why: 
Remote Issues needs to imported from a remote provider like github/jira and mapped into ALM as a ALM-friendly  workitem.

What:
- Add a static map to map remote work item fields into a local/ALM work item.
- Added a map/registry ( provider -> object ) to get the object based on provider name, which has the details implemented on how to 'fetch' the relevant fields from the remote work item ( which could be json/xml ).
- Added a `map` method which ( if considered a black box ) takes the static map and the remote work item as an input, and outputs an ALM persist-able WorkItem.




<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/almighty/almighty-core/152)
<!-- Reviewable:end -->

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,21 @@
+	}	
+	
+	for from_key, to_key := range wiMap {
+		workItem.Fields[to_key] = remoteWorkItem.Fields[from_key]
```
Check with @DhritiShikhar @Ritsyy , but from_key might need to be resolved as a child element of something. Don't remember how the github/jira issues look but as an example

```json
issue: {
  meta: {
    title: ""
  }
}
```

The map might need to be able to express: "issue/meta/title" -> "system.title". Depending on how the GitHub/Jira/Trello/Bugzilla structure is for the 3 fields of interested. If they are all Flat 1. level then we don't need it yet, if they are not we do.. :)

sbose78 replied:
Thanks @aslakknutsen .
This is how I plan to implement the multi-level key access:

The code which pulls the data from the remote provider would take care of flattening everything to level 1 ( I can't seem to think of a use case where this is not possible ).  

```
issue: {
  meta: {
    title: ""
  }
}
```  

should be sanitized/converted to a dict of :
```
{
   "title" : "some title",
  "description" : "some desc"
}
```
The flattened dict would be assigned to type `RemoteWorkItem`.

By doing the above, we can ensure that the mapping code doesn't change, and also that every provider will have it's own format so it should be best left to the integration code to standardize the 
structure into a Flat 1 evel.

aslakknutsen replied:
You're assuming there will only be 1 field named "title" in the whole doc if you flatten it like that. 

A flat v using the whole path "issue.meta.title" would work tho.

I think I would like to Fetch and Store the RemoteWorkItem as-is, but we can sure flatten it was a step after Store and Map.

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,29 @@
+        "title":"title"
+    }
+
+    workItem,err = remoteWorkItem.MapRemote(workItemMap,workItemType)    
```
Should probably verify the conversion as well :)

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,77 @@
+func (rwi *GithubRemoteWorkItem) GetWorkItemKeyMap() WorkItemMap {
+	workItemMap := WorkItemMap{
+		"id":    "remote_issue_id",
+		"body":  "description",
```
the to values should be: system.description, system.status, system.title

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,100 @@
+			Fields: make(map[string]interface{}),
+		},
+	}
+	githubRemoteWorkItem.Fields = responseJson
```
this could be set directly on line 60?

sbose78 replied:
Thanks, feedback incorporated.

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// GithubRemoteWorkItem is a derivative of RemoteWorkItem
+type GithubRemoteWorkItem struct {
```
Not quite sure why you have GithubRemoteWorkItem with a GetWorkItemKeyMap as oppose to:

```go
var remoteWorkItemMaps = map[string]WorkItemMap{"github":  WorkItemMap{....}}

workItem := MapRemoteWorkItem(remoteWorkItemMaps["github"], workitemtype)
 ```

sbose78 replied:
Thanks @aslakknutsen : 

I wanted to maintain the default key map at the ```GithubRemoteWorkItem``` implementation level and at the same time allow some flexibility to supply a different key maps at the ```MapRemoteWorkItem(..)```  invocation time, **also** because implementing the ```GetWorkItemKeyMap()``` method looks pretty optional. 

`var remoteWorkItemMaps = map[string]WorkItemMap{"github":  WorkItemMap{....}}`

I wouldn't want to do this because everytime we add  support for a new provider ( like Jira,bugzilla ), I wouldn't want a common data structure to be touched. ( unless we maintain a YAML/JSON file of maps )

aslakknutsen replied:
> I wouldn't want a common data structure to be touched

The structure is the same, you would only be adding values to the map.

As for now and some time, this will only be statically defined. If you want, hide the 'static map' behind;

```go

type RemoteWorkItemRepository struct {
  mappings map[string]WorkItemMap
}

func (repo *RemoteWorkItemRepository) Get(provider string, type WorkItemType) WorkItemMap {
  return repo.mappings[provider];
}

func NewRemoteWorkItemRepository() RemoteWorkItemRepository {
  return RemoteWorkItemRepository{
    mappings: map[string]WorkItemMap{"github": WorkItemMap{....}},
  }
}

```

aslakknutsen replied:
> I wanted to maintain the default key map at the GithubRemoteWorkItem implementation level and at the same time allow some flexibility to supply a different key maps at the MapRemoteWorkItem(..) invocation time, also because implementing the GetWorkItemKeyMap() method looks pretty optional.

The key at this stage is KISS. Our main concern should be the external API's. As certain ideas and concepts here are fairly undefined/unknown, let's keep it at a bare minimal. Just enough to fulfill the task at hand. Makes it so much easier to throw away later.

sbose78 commented:
@aslakknutsen 
I've incorporated all the feedback items mentioned. Please review.

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+// WorkItemMap : holds the mapping between Remote key and local key in a WI
+type WorkItemMap map[string]interface{}
```
This is the WorkItemMap, why map[string]interface{} as oppose to map[string]string ?

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,96 @@
+// Response from Provider API is sanitized into a
+// good looking flat 1-level dictionary
+type RemoteWorkItemMapper interface {
+	Flatten()
```
Shouldn't Flatten() return a flatten'ed map[string]interface{} ?

sbose78 replied:
No @aslakknutsen .
It's going to update the object member whose fields were an map[string]interface.

This is how it's being consumed:

```
githubRemoteWorkItem.Flatten()
```
Flatten() is actually optionally for XXXRemoteWorkItemType to implement. 
Example, in case of Github we already had everything **we need** as a flat 1-level dict

aslakknutsen replied:
How does the RemoteWorkItem know everything the WorkItemMap needs is on the 1 level?

Who is going to call this optional thing?

Who is making the decision to use GitHubRemoteWorkItem vs RemoteWorkItem?

sbose78 replied:
> How does the RemoteWorkItem know everything the WorkItemMap needs is on the 1 level?

It does not, and when a key is not found, I am planning to handle that as an error. but your concern is right.

> Who is going to call this optional thing?

The integration code. Can there be a miss? Yes. I agree, not a great design.

> Who is making the decision to use GitHubRemoteWorkItem vs RemoteWorkItem?

If for a provider a derived type does not exist, then the onus is on the integration code to flatten and then call the `MapRemote(..)` method. Else, the integration code can invoke a correct derivative of `RemoteWorkItem`. 

This is what I intended to do initially:

I would have wanted to have every derivative of  `RemoteWorkItem` override the`Flatten()` method, and at call the same from `func (rwi *RemoteWorkItem) MapRemote` .  
But I don't see how I can do that in Go.

sbose78 replied:
The other alternative would be to **not** have derivatives of `RemoteWorkItem` . The `MapRemote` method could conditionally decide how to flatten based on provider.


// MapRemote maps RemoteWorkItem to WorkItem
func (rwi *RemoteWorkItem) MapRemote(wiMap WorkItemMap, wiType *WorkItemType , Provider string ) WorkItem {
    //  Based on Provider call appropriate flattening method.
}
```

sbose78 replied:
@aslakknutsen ,
I was under an impression that derived types would need their own flattening method - but I guess the same flattening method should work across all providers ( as long as they *output* json ). ?

Should I go ahead with a single `Flatten` method belonging to `RemoteWorkItem` and call it inside `MapRemote` ?

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	for fromKey, toKey := range wiMap {
+		workItem.Fields[toKey.(string)] = rwi.Fields[fromKey]
```
No need for a cast if WorkItemMap is map[string]string

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	defer response.Body.Close()
+	responseData, err := ioutil.ReadAll(response.Body)
```
If you use https://golang.org/pkg/encoding/json/#NewDecoder you don't have to read the response data as a string first.

sbose78 replied:
Thanks,

Replaced it with 

```
responseDecoder := json.NewDecoder(response.Body)
err = responseDecoder.Decode(&responseJson)
```

sbose78 commented:
- Removed the derived types.
- added a single `Flatten` method which would be called inside the MapRemote() method.

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,92 @@
+		Fields: map[string]interface{}{
+			"id":          "3",
+			"description": "I am a remote issue's description in Github",
+			"title":       "I am the title of the issue",
```
Should hide one of the elements under a parent to verify that you can reach down into the format. e.g. 

```
{
 "id": "3",
 "meta": {
   "descrpition": "a",
   "title": "b"
 }
}
```

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+func TestMapGithub(t *testing.T) {
+	url := "https://api.github.com/repos/almighty/almighty-core/issues/131"
```
This depends on an external system we can't control which could make our test suite unstable for no good reason.

A good choice here would be to store the content locally and read it from a file instead. That of course, runs the risk of us not testing against real data if the remote system change in some form. 

A good strategy here would be to support 'automatic updatable test data examples'. Example:
https://gist.github.com/aslakknutsen/9ad87dcff9d6c5ad11584112d2deb5d6

sbose78 commented:
Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[models/remoteworkitem.go, line 59 \[r2\]](https://reviewable.io:443/reviews/almighty/almighty-core/152#-KPrdOPK84vlSdc1p7pl-r2-59:-KQ66g1nqkEEkklRiQQq:b-896fix) ([raw file](https://github.com/almighty/almighty-core/blob/3612048095bca2014cd0f3ae57e45bef0ad5af47/models/remoteworkitem.go#L59)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> the to values should be: system.description, system.status, system.title
</details>
Done.

---

*[models/remoteworkitem.go, line 18 \[r3\]](https://reviewable.io:443/reviews/almighty/almighty-core/152#-KPrdOPK84vlSdc1p7pl-r3-18:-KQ66W3m2p3wco_fDubj:b-r9dgwr) ([raw file](https://github.com/almighty/almighty-core/blob/f533bf96759bfe431d11322e0980975a717ee659/models/remoteworkitem.go#L18)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> This is the WorkItemMap, why map[string]interface{} as oppose to map[string]string ?
</details>
Incorporated.

---

*[models/remoteworkitem.go, line 93 \[r3\]](https://reviewable.io:443/reviews/almighty/almighty-core/152#-KPrdOPK84vlSdc1p7pl-r3-93:-KQ66hLUeL-8z0FvYlyr:b-896fix) ([raw file](https://github.com/almighty/almighty-core/blob/f533bf96759bfe431d11322e0980975a717ee659/models/remoteworkitem.go#L93)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> No need for a cast if WorkItemMap is map[string]string
</details>
Done.

---

*[models/remoteworkitem_test.go, line 27 \[r1\]](https://reviewable.io:443/reviews/almighty/almighty-core/152#-KPrdOPK84vlSdc1p7pm-r1-27:-KQ66wuPJEcj4Dx9ryTq:b-896fix) ([raw file](https://github.com/almighty/almighty-core/blob/eb4e62fccecbb14cb3cb758e8cc748fa14946fbe/models/remoteworkitem_test.go#L27)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Should probably verify the conversion as well :)
</details>
Done.

---

*[models/remoteworkitem_test.go, line 21 \[r3\]](https://reviewable.io:443/reviews/almighty/almighty-core/152#-KPrdOPK84vlSdc1p7pm-r3-21:-KQ66jxhmxEk9b8Q3oz_:b3x1hy4) ([raw file](https://github.com/almighty/almighty-core/blob/f533bf96759bfe431d11322e0980975a717ee659/models/remoteworkitem_test.go#L21)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> Should hide one of the elements under a parent to verify that you can reach down into the format. e.g. 
> 
> 
> 
> ```
> 
> {
> 
>  "id": "3",
> 
>  "meta": {
> 
>    "descrpition": "a",
> 
>    "title": "b"
> 
>  }
> 
> }
> 
> ``` 
</details>
Yet to add the Flatten() method. Github works fine without it.

---

*[models/remoteworkitem_test.go, line 42 \[r3\]](https://reviewable.io:443/reviews/almighty/almighty-core/152#-KPrdOPK84vlSdc1p7pm-r3-42:-KQ65sOq-JXE8M_xTmh9:bciwh5u) ([raw file](https://github.com/almighty/almighty-core/blob/f533bf96759bfe431d11322e0980975a717ee659/models/remoteworkitem_test.go#L42)):*
<details><summary><i>Previously, aslakknutsen (Aslak Knutsen) wrote…</i></summary>
> This depends on an external system we can't control which could make our test suite unstable for no good reason.
> 
> 
> 
> A good choice here would be to store the content locally and read it from a file instead. That of course, runs the risk of us not testing against real data if the remote system change in some form. 
> 
> 
> 
> A good strategy here would be to support 'automatic updatable test data examples'. Example:
> 
> https://gist.github.com/aslakknutsen/9ad87dcff9d6c5ad11584112d2deb5d6
> 
> 
> 
> 
</details>
Incorporated these.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/almighty/almighty-core/152)*
<!-- Sent from Reviewable.io -->

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,67 @@
+	Content []byte
+}
+
+// GitHubRemoteWorkItem knows how to implement a FieldAccessor on a GitHub Issue JSON struct
```
Update doc to use AttributeAccesor

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,67 @@
+	issue map[string]interface{}
+}
+
+// NewGitHubRemoteWorkItem creates a new Decoded FieldAccessor for a GitHub Issue
```
Update doc to use AttributeAccesor

sbose78 replied:
Sure, Btw, Not using FieldExpression and using AttributeExpression because it's already being used in the `criteria` package.

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,116 @@
+// Data is stored under examples/test
+// This is done to avoid always depending on remote systems, but also with an option
+// to refresh/retest against the 'current' remote system data without manual copy/paste
+func LoadTestData(filename string, provider TestDataProvider) ([]byte, error) {
```
This should be extracted out into a separate file.

sbose78 replied:
Thanks, extracted out.

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// RemoteWorkItem is the Database stored TrackerItem
+type RemoteWorkItem struct {
```
This object is technically provided by https://github.com/almighty/almighty-core/pull/111

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,67 @@
+	SystemTitle         = "system.title"
+	SystemDescription   = "system.description"
+	SystemStatus        = "system.status"
+	ProviderGithub      = "github"
```
This object is technically provided by #111

sbose78 replied:
Thanks.

Hi @baijum FYI.
This `const` declaration holds the key of the `trackerTypes` map in models/tracker_respository.go

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,116 @@
+
+func TestGitHubIssueMapping(t *testing.T) {
+
+	content, err := LoadTestData("github_issue_mapping.json", provideRemoteGithubData)
```
semi-point would be to also commit the github_issue_mapping.json file :)

sbose78 replied:
Thanks, Committed `models/examples/test/githubi_issue_mapping.json`

aslakknutsen commented on models/remoteworkitem_test.go:
```diff
@@ -0,0 +1,116 @@
+	}
+	remoteWorkItem := RemoteWorkItem{ID: "xyz", Content: []byte("{\"title\":\"abc\"}")}
+
+	gh, err := NewGitHubRemoteWorkItem(remoteWorkItem)
```
A little registry that maps ProviderGithub to the function NewGitHubRemoteWorkItem for lookup by the 'caller'

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+const (
+	SystemRemoteIssueId = "system.remote_issue_id"
```
I think #111 is aiming for this to be a URL, but handled as a unique ID.

sbose78 replied:
OK, in that case I think we should keep the name as it is ( "remote_issue_id" ) since it's being handled as a unique ID , just to keep things broad ?
Or do you suggest I update this to "remote_issue_url" ?

aslakknutsen replied:
Not sure, was just a FYI :)

aslakknutsen commented on models/remoteworkitem.go:
```diff
@@ -0,0 +1,67 @@
+
+// Map maps the remote WorkItem to a local WorkItem
+func Map(item AttributeAccesor, mapping WorkItemMap) (WorkItem, error) {
+	workItem := WorkItem{Fields: Fields{}}
```
The WorkItem here should be app.WorkItem btw. As that's what the WorkItemRepository expects.

sbose78 commented:
@aslakknutsen 
Made all the changes proposed. Could you please review?

aslakknutsen commented on models/remote_test_data.go:
```diff
@@ -0,0 +1,48 @@
+package models
```
This should probably be in test/remote_data_test.go in package test

sbose78 replied:
Placed it inside 
`src/github.com/almighty/almighty-core/test/remote_test_data.go`

baijum commented on remoteworkitem/remoteworkitem_test.go:
```diff
@@ -0,0 +1,80 @@
+	workItemMap := WorkItemMap{
+		AttributeExpression("title"): "system.title",
+	}
+	remoteWorkItem := RemoteWorkItem{ID: "xyz", Content: []byte("{\"title\":\"abc\"}")}
```
You can use back quote for string literals like this:

    `{"title":"abc"}`

baijum commented on examples/test/github_issue_mapping.json:
```diff
@@ -0,0 +1 @@
+{"url":"https://api.github.com/repos/almighty/almighty-core/issues/131","repository_url":"https://api.github.com/repos/almighty/almighty-core","labels_url":"https://api.github.com/repos/almighty/almighty-core/issues/131/labels{/name}","comments_url":"https://api.github.com/repos/almighty/almighty-core/issues/131/comments","events_url":"https://api.github.com/repos/almighty/almighty-core/issues/13 ...
```
Do we need the `test` directory under `examples`. Can we keep it directly under `examples`?

aslakknutsen replied:
Maybe the examples parent directory is the wrong place in general. The examples directory is used to provide examples of 'our' API payloads as used from CLI.

These new files are stored response data from external services for test purposes. 

test_data ?

sbose78 replied:
The `test` directory would kind of differentiate the file from other examples in a way that it would imply that the code consumes the file and hence should not be touched without reason :)

baijum replied:
How about moving it under `test/data`?

sbose78 replied:
I agree @baijum and @aslakknutsen  
I think test/data would make more sense.

aslakknutsen replied:
Sounds good

baijum commented on remoteworkitem/remoteworkitem.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+const (
+	SystemRemoteIssueId = "system.remote_issue_id"
```
The word "issue" is not commonly used by all remote trackers. I think we can replace it with "item".

sbose78 replied:
Very good find. Will change this @baijum .

baijum commented on remoteworkitem/remoteworkitem_test.go:
```diff
@@ -0,0 +1,80 @@
+
+func provideRemoteGithubData() ([]byte, error) {
+	url := "https://api.github.com/repos/almighty/almighty-core/issues/131"
+	response, err := http.Get(url)
```
I think "unit" test should not make external HTTP requests. Can we use a mock object?

sbose78 replied:
You are correct. This method is called optionally if the example/test/<locally_stored_output>.json is not present or if a flag is passed explicitly for refreshing the xample/test/<locally_stored_output>.json 

Otherwise this is not called.

baijum commented on test/remote_test_data.go:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	// The target dir would be src/github.com/almighty/almighty-core/examples/test
+	targetDir := path.Dir(packagefilename) + "/../examples/test/"
```
This code will not work under Microsoft Windows as the path separator is Unix style.

sbose78 replied:
Thanks.
Fixed this to use 
`filepath.ToSlash(path string)`

baijum commented on remoteworkitem/remoteworkitem_test.go:
```diff
@@ -0,0 +1,80 @@
+		t.Error("system.title not mapped")
+	}
+
+	fmt.Println(workItem)
```
use `t.Log(workItem)` ref. https://golang.org/pkg/testing/#T.Log

baijum commented on remoteworkitem/remoteworkitem_test.go:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	for _, localWorkItemKey := range workItemMap {
+		fmt.Println("Mapped ", localWorkItemKey)
```
use `t.Log(workItem)` ref. https://golang.org/pkg/testing/#T.Log

sbose78 replied:
Changes incorporated.

aslakknutsen commented:
:Looks good to me. @sbose78 Could you create the following 'sub-tasks' to be handled either as part of #111 or when #111 comes in:

Align Provider names with #111 https://github.com/almighty/almighty-core/pull/152/files#diff-b9d69c203555944a118b1e5b55713cc1R14

Move Field Type names to WorkItem 'somewhere' https://github.com/almighty/almighty-core/pull/152/files#diff-b9d69c203555944a118b1e5b55713cc1R10

Replace RemoteWorkItem with RemoteWorkItem from #111 https://github.com/almighty/almighty-core/pull/152/files#diff-b9d69c203555944a118b1e5b55713cc1R38
